#4034 Confirmation dialog in map on user logout

kboczkowski-clurgo opened the pull request:
## Description
In this PR I've added before leave popup with map changes validation

## Issues
 - #4034 
 - ...

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** (check one with "x", remove the others)

 - [ ] Bugfix
 - [x] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:


**What is the current behavior?** (You can also link to an open issue here)


**What is the new behavior?**


**Does this PR introduce a breaking change?** (check one with "x", remove the other)

 - [ ] Yes, and I documented them in migration notes
 - [ ] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

MV88 commented on web/client/localConfig.json:
```diff
@@ -681,5 +681,6 @@
         "manager": ["Header", "Redirect", "Manager", "Home", "UserManager", "GroupManager", "Footer"]
-    }
+    },
+    "unsavedMapChangesDialog": true
```
Please do not change localConfig when not needed.  
default values should be used elsewhere (default prop or in selector)
localConfig is used to override/customize default behaviors

- please add some documentation to the local-config.md for this new config property

MV88 commented on web/client/translations/data.it-IT:
```diff
@@ -310,7 +310,11 @@
-                "createNewOne": "Creane una nuova"
+                "createNewOne": "Creane una nuova",
+                "unsavedMapConfirmTitle": "Sei sicuro",
+                "unsavedMapConfirmMessage": "Sei sicuro di voler lasciare modifiche non salvate?",
```
```suggestion
                "unsavedMapConfirmMessage": "Sei sicuro di voler uscire senza salvare le modifiche?",
```

MV88 commented on web/client/translations/data.it-IT:
```diff
@@ -310,7 +310,11 @@
+                "createNewOne": "Creane una nuova",
+                "unsavedMapConfirmTitle": "Sei sicuro",
+                "unsavedMapConfirmMessage": "Sei sicuro di voler lasciare modifiche non salvate?",
+                "unsavedMapConfirmButtonText": "Partire",
```
```suggestion
                "unsavedMapConfirmButtonText": "Si",
```

MV88 commented on web/client/translations/data.it-IT:
```diff
@@ -310,7 +310,11 @@
+                "unsavedMapConfirmTitle": "Sei sicuro",
+                "unsavedMapConfirmMessage": "Sei sicuro di voler lasciare modifiche non salvate?",
+                "unsavedMapConfirmButtonText": "Partire",
+                "unsavedMapCancelButtonText": "Chiudi"
```
```suggestion
                "unsavedMapCancelButtonText": "No"
```

MV88 commented on web/client/translations/data.en-US:
```diff
@@ -310,7 +310,11 @@
                 "noMapAvailable": "No map available",
-                "createNewOne": "Create a new one"
+                "createNewOne": "Create a new one",
+                "unsavedMapConfirmTitle": "Are you sure",
```
i would use something like 
```suggestion
                "unsavedMapConfirmTitle": "Unsaved changes",
```

please update others titles

MV88 commented on web/client/utils/MapUtils.js:
```diff
@@ -414,6 +414,20 @@ const parseLayoutValue = (value, size = 0) => {
+        'widgetsConfig'
+    ];
+
+    return JSON.stringify(pick(originalState, pickedFields)) === JSON.stringify(pick(newState, pickedFields));
```
i would use IsEqual of lodash here, instead of === 

what do you think?

MV88 commented on web/client/utils/__tests__/MapUtils-test.js:
```diff
@@ -1931,4 +1932,202 @@ describe('Test the MapUtils', () => {
+            "widgetsConfig": {},
+            "mapInfoConfiguration": {}
+        };
+        expect(compareMapChanges(map1, map2)).toBeFalsy();
```
Hi, this should be true instead of false. (the title of the test is suggesting this)

MV88 commented on web/client/epics/map.js:
```diff
@@ -214,11 +222,48 @@ const checkMapPermissions = (action$, {getState = () => {} }) =>
+            return Rx.Observable.defer(() => axios.get(url))
+                .switchMap(({ data: originalMap }) => {
+                    const updatedMap = MapUtils.saveMapConfiguration(
+                        mapSelector(getState()),
```
please, use a constant to store the  for getState() instead of many function calls

kboczkowski-clurgo replied:
done

MV88 commented on web/client/epics/map.js:
```diff
@@ -214,11 +222,48 @@ const checkMapPermissions = (action$, {getState = () => {} }) =>
+            const { view } = getState().feedbackMask;
+
+            if (!mapId && view !== 'viewer') {
+                action && action();
```
i will grab this from you, really nice

MV88 commented on web/client/epics/map.js:
```diff
@@ -214,11 +222,48 @@ const checkMapPermissions = (action$, {getState = () => {} }) =>
+                            setControlProperty('unsavedMap', 'source', source, false)
+                        );
+                    }
+                    action && action();
```
Could this be better?

return action ? Rx.Observable.of(action) : Rx.Observable.empty();

kboczkowski-clurgo replied:
agree i fixed it

kboczkowski-clurgo commented:
I've updated code, so @MV88 you can take a look on that now

MV88 commented on web/client/epics/map.js:
```diff
@@ -214,11 +222,55 @@ const checkMapPermissions = (action$, {getState = () => {} }) =>
+
+            if (mapId) {
+                const url = ConfigUtils.getConfigUrl({ mapId }).configUrl;
+                return Rx.Observable.defer(() => axios.get(url))
```
would be possible to take originalMap when doing the first request to get the map info ?

kboczkowski-clurgo replied:
Yes, i've done it now :)

kboczkowski-clurgo commented:
I've fixed problem with roles, but i also found an issue:
![Zrzut ekranu z 2019-11-18 14-29-38](https://user-images.githubusercontent.com/56867851/69056310-f89c4700-0a0f-11ea-9c5d-3c9d283fedb8.png)

Here in map configuration object fetched from api there are fields like `map.backgrounds` and `widgetsConfig.layouts.md` which are arrays. Updated data contains those fields but they are objects. I have a problem now, because there are many tests on frontend, which are based on that case (objects). @MV88 what do you think about that, maybe we can adjust those data to be objects from the API perspective?

MV88 commented:
> I've fixed problem with roles, but i also found an issue:
> ![Zrzut ekranu z 2019-11-18 14-29-38](https://user-images.githubusercontent.com/56867851/69056310-f89c4700-0a0f-11ea-9c5d-3c9d283fedb8.png)
> 
> Here in map configuration object fetched from api there are fields like `map.backgrounds` and `widgetsConfig.layouts.md` which are arrays. Updated data contains those fields but they are objects. I have a problem now, because there are many tests on frontend, which are based on that case (objects). @MV88 what do you think about that, maybe we can adjust those data to be objects from the API perspective?

i think the contrary, to get arrays instead of objects,  
where they becomes object?
i see that layers are arrays
![image](https://user-images.githubusercontent.com/11991428/69537743-078b7800-0f81-11ea-8c73-a1c52d4bd599.png)

MV88 commented:
please solve the conflict

MV88 commented:
also have a look at the compare function since in the updated map there is a typo
![image](https://user-images.githubusercontent.com/11991428/69538027-aadc8d00-0f81-11ea-858e-c720a3b40741.png)
and here 
https://github.com/geosolutions-it/MapStore2/pull/4486/files#diff-46f38f1c08643e9486be7d5de65c2c12R422

kboczkowski-clurgo commented:
> also have a look at the compare function since in the updated map there is a typo
> ![image](https://user-images.githubusercontent.com/11991428/69538027-aadc8d00-0f81-11ea-858e-c720a3b40741.png)
> and here
> https://github.com/geosolutions-it/MapStore2/pull/4486/files#diff-46f38f1c08643e9486be7d5de65c2c12R422

I fixed typo and resolved conflicts, now should work ok @MV88

MV88 commented:
there is a problem with commits and file changed here

@kboczkowski-clurgo  can you fix them?

kboczkowski-clurgo commented:
Closed, and moved here: https://github.com/geosolutions-it/MapStore2/pull/4581
